oci/archive/oci_src.go

```diff
@@ -0,0 +1,89 @@
+package archive
+
+import (
+	"context"
+	"io"
+
+	ocilayout "github.com/containers/image/oci/layout"
+	"github.com/containers/image/types"
+	digest "github.com/opencontainers/go-digest"
+	imgspecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+type ociArchiveImageSource struct {
+	ref         ociArchiveReference
+	unpackedSrc types.ImageSource
+	tempDir     string
+}
+
+// newImageSource returns an ImageSource for reading from an existing directory.
+// newImageSource untars the file and saves it in a temp directory
+func newImageSource(ref ociArchiveReference, ctx *types.SystemContext, requestedManifestMIMETypes []string) (types.ImageSource, error) {
+	if err := createUntarTempDir(&ref); err != nil {
+		return nil, errors.Wrap(err, "error creating temp directory")
+	}
+
+	unpackedSrc, err := ref.tempDirRef.ociRefExtracted.NewImageSource(ctx, requestedManifestMIMETypes)
+	if err != nil {
+		if err := ref.tempDirRef.deleteTempDir(); err != nil {
+			return nil, errors.Wrapf(err, "error deleting temp directory", ref.tempDirRef.tempDirectory)
+		}
+		return nil, err
+	}
+	return &ociArchiveImageSource{ref: ref,
+		unpackedSrc: unpackedSrc,
+		tempDir:     ref.tempDirRef.tempDirectory}, nil
+}
+
+// LoadManifestDescriptor loads the manifest
+func LoadManifestDescriptor(imgRef types.ImageReference) (imgspecv1.Descriptor, error) {
+	ociArchRef, ok := imgRef.(ociArchiveReference)
+	if !ok {
+		return imgspecv1.Descriptor{}, errors.Errorf("error typecasting, need type ociArchiveReference")
+	}
+	if err := createUntarTempDir(&ociArchRef); err != nil {
+		return imgspecv1.Descriptor{}, errors.Wrap(err, "error creating temp directory")
+	}
+	defer ociArchRef.tempDirRef.deleteTempDir()
+
+	descriptor, err := ocilayout.LoadManifestDescriptor(ociArchRef.tempDirRef.ociRefExtracted)
+	if err != nil {
+		if err := ociArchRef.tempDirRef.deleteTempDir(); err != nil {
```

Comment:
There already is a `defer ….deleteTempDir()` above; and deleting the same name from a shared `*/tmp` directory twice is not, strictly speaking, safe (the name could have been reused by an unrelated process in the meantime).

If the point is to not ignore the error, that can be done in the `defer` block, e.g. see how `copy.Image` does that—but only if it is really important to preserve the cleanup error.

In this particular case it seems that the `error loading index` error is more important than the `error deleting temporary directory`, at least the `error loading index` one shouldn't be thrown away; though notifying the user that the cleanup failed might also be useful, it is less important. (See also the discussion around https://github.com/containers/image/pull/236#discussion_r100569280 if you care.)

Reply:
Since the `error loading index` error is more important, I didn't do what `copy.Image` does. Just took out the second `deleteTempDir()` and left it as a defer.